PopPUNK/network.py

```diff
@@ -704,3 +724,112 @@ def printExternalClusters(newClusters, extClusterFile, outPrefix,
+        weighted_network = True
+    # Calculate minimum spanning tree
+    if weighted_network:
+        mst_edge_prop_map = gt.min_spanning_tree(G, weights = G.ep["weight"])
+    else:
+        mst_edge_prop_map = gt.min_spanning_tree(G)
+    mst_network = gt.GraphView(G, efilt = mst_edge_prop_map)
+    # Find seed nodes as those with greatest outdegree in each component
+    seed_vertices = set()
+    component_assignments, component_frequencies = gt.label_components(mst_network)
+    for component_index in range(len(component_frequencies)):
+        component_members = component_assignments.a == component_index
+        component = gt.GraphView(mst_network, vfilt = component_members)
+        component_vertices = component.get_vertices()
+        out_degrees = component.get_out_degrees(component_vertices)
+        seed_vertex = list(component_vertices[np.where(out_degrees == np.amax(out_degrees))])
+        seed_vertices.add(seed_vertex[0]) # Can only add one otherwise not MST
+    # If multiple components, calculate distances between seed nodes
+    if len(component_frequencies) > 1:
+        # Extract distances
+        connections = []
+        for row_idx, (ref, query) in enumerate(listDistInts(names, names, self = True)):
+            if ref in seed_vertices and query in seed_vertices:
+                dist = distMat[row_idx, 0]
+                edge_tuple = (ref, query, dist)
+                connections.append(edge_tuple)
+        # Construct graph
+        seed_G = gt.Graph(directed = False)
+        seed_G.add_vertex(len(seed_vertex))
+        if weighted_network:
+            eweight = seed_G.new_ep("float")
+            seed_G.add_edge_list(connections, eprops = [eweight])
+            seed_G.edge_properties["weight"] = eweight
+            seed_mst_edge_prop_map = gt.min_spanning_tree(seed_G, weights = seed_G.ep["weight"])
+            seed_mst_network = gt.GraphView(seed_G, efilt = seed_mst_edge_prop_map)
+            # Insert seed MST into original MST - may be possible to use graph_union with include=True & intersection
+            deep_edges = seed_mst_network.get_edges([seed_mst_network.ep["weight"]])
+            mst_network.add_edge_list(deep_edges)
+        else:
+            sys.stderr.write('I do not currently know what to do in this case\n')
+    # Initialise tree
+    tree = dendropy.Tree(taxon_namespace=taxon_namespace)
+    # Create nodes
+    tree_nodes = {v:dendropy.Node(taxon=taxon_namespace[int(v)]) for v in mst_network.get_vertices()}
+    # Identify edges
+    tree_edges = {v:[] for v in tree_nodes.keys()}
+    tree_edge_lengths = {v:[] for v in tree_nodes.keys()}
+    if weighted_network:
+        network_edge_weights = list(mst_network.ep["weight"])
+    else:
+        network_edge_weights = [1.0]*len(mst_network.get_edges())
+    for i,edge in enumerate(mst_network.get_edges()):
+        # Connectivity
+        tree_edges[edge[0]].append(edge[1])
+        tree_edges[edge[1]].append(edge[0])
```

Comment:
This adds a self loop?
Generally a bit unsure what this part of the code is doing, so maybe some more comments would be helpful

Reply:
Added explanation here:
`        # Connectivity - add both directions as unrooted tree is not directional -
        # do not know which will be leaf node
        tree_edges[edge[0]].append(edge[1])
        tree_edges[edge[1]].append(edge[0])
        # Lengths added in the same order as the corresponding children to enable
        # branches to be matched to child nodes
        tree_edge_lengths[edge[0]].append(network_edge_weights[i])
        tree_edge_lengths[edge[1]].append(network_edge_weights[i])`
        
In commit 9bc521f.